### PR TITLE
make Mingw build work on Windows

### PR DESCRIPTION
**Pull request by eiennohito**
MSVC also builds, but haven't checked tests

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -15,10 +15,15 @@ endif()
 set(CMAKE_CXX_STANDARD 14)
 #Ensures MSVC treats strings as UTF-8 at both compile/runtime time
 if (MSVC)
+	add_compile_options("/D_WIN32_WINNT=0x0600" "/wd4141" "/wd4018")
```
Comment (DoumanAsh) [inline CMakeLists.txt]: Should `_WIN32_WINNT` defined through `if(WIN32)` while specific flags with the same `if(MSVC)`?

File: src/util/mmap_impl_win32.h [hunk 2]
```diff
@@ -155,16 +153,15 @@ Status MappedFile::map(MappedFileFragment *view, size_t offset, size_t size,
                               << ". Error: " << GetLastError();
   }
 
-  const DWORD granuality = get_system_granuality();
+  DWORD granuality = get_system_granuality();
```
Comment (DoumanAsh) [inline src/util/mmap_impl_win32.h]: Shouldn't it remain const?, it is not supposed to be changed further in code

**Comment by DoumanAsh** [comment]
I will check later on how MSVC works there and will make necessary adjustments

File: CMakeLists.txt (hunk 1 above)
Comment (eiennohito) [reply CMakeLists.txt]: /D vs -D problems :p

Comment (DoumanAsh) [reply CMakeLists.txt]: you can use `add_definitions` where CMake handles it but well i suppose these IFs are nearby so it is not a problem

File: src/util/mmap_impl_win32.h (hunk 2 above)
Comment (eiennohito) [reply src/util/mmap_impl_win32.h]: I try not to reuse variables as much time as it is possible.
At the same time, other code does not use const in local functions as well.

const T& vs T& for parameters is very important on the other hand.

File: src/util/mmap_impl_win32.h [hunk 3]
```diff
@@ -101,15 +90,23 @@ inline static DWORD get_system_granuality() {
...
-                                                        sizeof(info)));
+  BY_HANDLE_FILE_INFORMATION fileInfo = {0};
+  if (GetFileInformationByHandle(handle, &fileInfo) == 0) {
+    return false;
+  }
+
+  LARGE_INTEGER info_size;
+  info_size.LowPart = fileInfo.nFileSizeLow;
+  info_size.HighPart = fileInfo.nFileSizeHigh;
+  if (info_size.QuadPart >= size) {
```
Comment (DoumanAsh) [inline src/util/mmap_impl_win32.h]: Btw there is another check that check out whether we extend above existing file size using `size_` member
Should it be then removed there?

File: src/util/mmap_impl_win32.h (hunk 2 above)
Comment (eiennohito) [reply src/util/mmap_impl_win32.h]: Btw, the result of the get_system_granlarity call was not used in your patch and I didn't see it.

File: src/util/mmap_impl_win32.h (hunk 3 above)
Comment (eiennohito) [reply src/util/mmap_impl_win32.h]: In the current impl, it's possible to remove this and still be correct, yep.

File: src/util/mmap_impl_win32.h (hunk 2 above)
Comment (DoumanAsh) [reply src/util/mmap_impl_win32.h]: Uh... looking at code yeah... It seems I moved result of function to variable, but forgot to remove invocation https://github.com/ku-nlp/jumanpp/commit/0be5fa004d145f05c07f5185effa697403fda454#diff-ebcfd21b2ea62c81d687e6e35fdc99aeR158

Really lame of me

Comment (DoumanAsh) [reply src/util/mmap_impl_win32.h]: To be honest Im not really sure it is good idea to re-use variable as much as possible.
Minimizing variable scope and making them const makes it easier to reason about code which consequently makes it easier to understand code.

But this is just my opinion, if you'd like to keep it like that it is ok too.

Comment (eiennohito) [reply src/util/mmap_impl_win32.h]: **not to reuse**.
I agree that reusing should be done only when its needed.
Register allocators are pretty smart to get the score right.

It's just... the other code here doesn't use const for locals.

**Pull request merged**